reference/reorder/mc64_kernels.cpp

```diff
@@ -0,0 +1,440 @@
+                        const matrix::Csr<ValueType, IndexType>* mtx,
+                        array<remove_complex<ValueType>>& value_workspace,
+                        gko::reorder::reordering_strategy strategy)
+{
+    constexpr auto inf =
+        std::numeric_limits<remove_complex<ValueType>>::infinity();
+    const auto nnz = mtx->get_num_stored_elements();
+    const auto num_rows = mtx->get_size()[0];
+    const auto row_ptrs = mtx->get_const_row_ptrs();
+    const auto col_idxs = mtx->get_const_col_idxs();
+    const auto values = mtx->get_const_values();
+    auto calculate_weight =
+        strategy == gko::reorder::reordering_strategy::max_diagonal_sum
+            ? [](ValueType a) { return abs(a); }
+            : [](ValueType a) { return std::log2(abs(a)); };
+    auto weights = value_workspace.get_data();
+    auto dual_u = weights + nnz;
+    auto distance = dual_u + num_rows;
+    auto row_maxima = distance + num_rows;
+    for (IndexType col = 0; col < num_rows; col++) {
+        dual_u[col] = inf;
+        distance[col] = inf;
+    }
+
+    for (IndexType row = 0; row < num_rows; row++) {
+        const auto row_begin = row_ptrs[row];
+        const auto row_end = row_ptrs[row + 1];
+        auto row_max = -inf;
+        for (IndexType idx = row_begin; idx < row_end; idx++) {
+            const auto weight = calculate_weight(values[idx]);
+            weights[idx] = weight;
+            if (weight > row_max) row_max = weight;
+        }
+
+        row_maxima[row] = row_max;
+
+        for (IndexType idx = row_begin; idx < row_end; idx++) {
+            const auto weight = row_max - weights[idx];
+            weights[idx] = weight;
+            const auto col = col_idxs[idx];
+            if (weight < dual_u[col]) dual_u[col] = weight;
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_MC64_INITIALIZE_WEIGHTS_KERNEL);
+
+
+// Assume -1 in permutation and inv_permutation
+template <typename ValueType, typename IndexType>
+void initial_matching(std::shared_ptr<const DefaultExecutor> exec,
+                      size_type num_rows, const IndexType* row_ptrs,
+                      const IndexType* col_idxs,
+                      const array<ValueType>& value_workspace,
+                      array<IndexType>& permutation,
+                      array<IndexType>& inv_permutation,
+                      array<IndexType>& index_workspace, ValueType tolerance)
+{
+    const auto nnz = row_ptrs[num_rows];
+    const auto weights = value_workspace.get_const_data();
+    const auto dual_u = weights + nnz;
+    auto p = permutation.get_data();
+    auto ip = inv_permutation.get_data();
+    auto idxs = index_workspace.get_data() + 4 * num_rows;
+    auto unmatched = idxs + num_rows;
+    auto um_cnt = 0;
+
+    // For each row, look for an unmatched column col for which weight(row, col)
+    // = 0. If one is found, add the edge (row, col) to the matching and move on
```

Comment:
This comment seems to be outdated/incorrect. Since the requirement is to be < tolerance and ip[col] == -1. Also what is dual_u, seems to be undocumented.

Reply:
The tolerance here is needed since logarithm computations have enough rounding errors to screw up exact equality in a lot of cases. ip[col] == -1 means the column is unmatched. I add a comment on this.